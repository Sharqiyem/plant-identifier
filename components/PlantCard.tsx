import React from 'react';
import { View, Text, Platform } from 'react-native';
import { Plant } from '@/lib/types';

interface PlantCardProps {
  plant: Plant;
  showTimestamp?: boolean;
  timestamp?: string;
}

export const PlantCard: React.FC<PlantCardProps> = ({
  plant,
  showTimestamp = false,
  timestamp
}) => {
  const shadowStyle = Platform.select({
    ios: 'shadow-lg shadow-black/50',
    android: 'elevation-4'
  });

  return (
    <View className={`bg-card py-4 rounded-lg mb-4 ${shadowStyle}`}>
      <Text className="text-card-foreground text-2xl font-bold mb-2">Name: {plant.name}</Text>
      <Text className="text-muted-foreground mb-1">Scientific Name: {plant.scientificName}</Text>
      <Text className="text-muted-foreground mb-1">Family: {plant.family}</Text>
      <Text className="text-foreground mb-2">{plant.description}</Text>
      {showTimestamp && timestamp && (
        <Text className="text-muted-foreground text-sm">
          {new Date(timestamp).toLocaleString()}
        </Text>
      )}
    </View>
  );
};
