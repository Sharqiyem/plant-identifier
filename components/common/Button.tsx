import React from 'react';
import { TouchableOpacity, Text } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { twMerge } from 'tailwind-merge';
import Colors from '@/constants/Colors';

interface ButtonProps {
  title: string;
  iconName: keyof typeof Ionicons.glyphMap;
  onPress: () => void;
  disabled?: boolean;
  className?: string;
}

export const Button: React.FC<ButtonProps> = ({
  title,
  iconName,
  onPress,
  className,
  disabled,
  ...props
}) => (
  <TouchableOpacity
    testID="button"
    className={twMerge(
      'bg-primary rounded-lg px-6 py-3 flex-row justify-center items-center',
      className
    )}
    onPress={onPress}
    disabled={disabled}
    {...props}
  >
    <Ionicons testID="icon" name={iconName} size={22} color={Colors.text} />
    <Text className="text-foreground font-bold ml-2">{title}</Text>
  </TouchableOpacity>
);
