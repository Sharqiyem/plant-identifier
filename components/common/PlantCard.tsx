import React from 'react';
import { View, Text, Platform } from 'react-native';
import { Plant } from '@/types';

interface PlantCardProps {
  plant: Plant;
  showTimestamp?: boolean;
  timestamp?: string;
}

export const PlantCard: React.FC<PlantCardProps> = ({
  plant,
  showTimestamp = false,
  timestamp
}) => {
  const shadowStyle = Platform.select({
    ios: 'shadow-lg shadow-black/50',
    android: 'elevation-4'
  });
  return (
    <View className={`bg-card py-4 rounded-lg mb-4 ${shadowStyle}`}>
      <Text testID="plant-name" className="text-card-foreground text-2xl font-bold mb-2">
        Name: {plant.name}
      </Text>
      <Text testID="plant-scientific-name" className="text-muted-foreground mb-1">
        Scientific Name: {plant.scientificName}
      </Text>
      <Text testID="plant-family" className="text-muted-foreground mb-1">
        Family: {plant.family}
      </Text>
      <Text testID="plant-description" className="text-foreground mb-2" numberOfLines={3}>
        {plant.description}
      </Text>
      {showTimestamp && timestamp && (
        <Text testID="plant-timestamp" className="text-muted-foreground text-sm">
          {new Date(timestamp).toLocaleString()}
        </Text>
      )}
    </View>
  );
};

// import React from 'react';
// import { View, Text, Platform } from 'react-native';
// import { Plant } from '@/types';

// export const PlantCard = () => {
//   return (
//     <View className="p-4 bg-card rounded-lg shadow-md">
//       <View className="flex-row items-center justify-center mb-3">
//         {/* <Ionicons name="information-circle-outline" color={Colors.primary} size={24} /> */}
//         <Text className="ml-2 text-lg font-semibold text-primary">How to Use</Text>
//       </View>
//       <Text className="text-center text-base text-foreground leading-6">
//         To identify plants, select an image from your gallery or take a photo using the camera. Our
//         AI will analyze the image and provide detailed information about the plant.
//       </Text>
//     </View>
//   );
// };
