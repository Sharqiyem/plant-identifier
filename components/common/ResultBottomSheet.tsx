import Colors from '@/constants/Colors';
import { Language } from '@/types';
import {
  BottomSheetFlatList,
  BottomSheetModal,
  BottomSheetModalProvider
} from '@gorhom/bottom-sheet';
import { useRouter } from 'expo-router';
import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useState } from 'react';
import { Image, Text, View } from 'react-native';
import { PlantCard } from './PlantCard';

type ResultBottomSheetProps = {
  plantInfo: any;
  languages: any;
};

export type ResultBottomSheetRef = {
  present: () => void;
};

const ResultBottomSheet = forwardRef<ResultBottomSheetRef, ResultBottomSheetProps>(
  ({ plantInfo, languages }, ref) => {
    const bottomSheetModalRef = React.useRef<BottomSheetModal>(null);
    const router = useRouter();

    const [selectedLanguage] = useState(languages[0]?.languageCode || 'en');

    const languageCodes = Object.keys(plantInfo);
    const previewUri = plantInfo[languageCodes[0]]?.previewUri;

    // variables
    const snapPoints = useMemo(() => ['95%'], []);

    // callbacks
    const handleSheetChanges = useCallback(
      (index: number) => {
        console.log('handleSheetChanges', index);
        if (index === -1) {
          // Sheet is closed
          router.back();
        }
      },
      [router]
    );

    useImperativeHandle(ref, () => ({
      present: () => {
        bottomSheetModalRef.current?.present();
      }
    }));

    // renders
    return (
      <BottomSheetModalProvider>
        <BottomSheetModal
          ref={bottomSheetModalRef}
          index={0}
          snapPoints={snapPoints}
          onChange={handleSheetChanges}
          enablePanDownToClose={true}
          handleStyle={{ backgroundColor: Colors.background }}
          handleComponent={() => (
            <View className="rounded-t-xl h-4  border-none justify-center items-center bg-background">
              <View className="w-14 h-[6px] rounded-full bg-primary" />
            </View>
          )}
        >
          <View className="flex-1 p-4 bg-background">
            <Text className="text-foreground text-2xl font-bold mb-4 ">
              Plant Identification Results
            </Text>
            {previewUri && (
              <Image
                testID="plant-image"
                source={{ uri: previewUri }}
                className="w-full h-48 object-cover"
              />
            )}

            {plantInfo[selectedLanguage] && (
              <View className="px-0 ">
                <PlantCard testID="plant-card" plant={plantInfo[selectedLanguage]} />
              </View>
            )}

            <Text className="text-lg font-semibold text-primary mt-4 mb-2">All Languages</Text>

            <BottomSheetFlatList
              showsVerticalScrollIndicator={false}
              data={languageCodes}
              renderItem={({ item }) => (
                <View className="mb-6  ">
                  <Text className="text-lg font-semibold text-accent mb-2">
                    {languages.find((lang: Language) => lang.languageCode === item)?.languageName ||
                      item}
                  </Text>
                  <PlantCard testID="plant-card" plant={plantInfo[item]} />
                </View>
              )}
              keyExtractor={(item) => item}
            />
          </View>
        </BottomSheetModal>
      </BottomSheetModalProvider>
    );
  }
);

ResultBottomSheet.displayName = 'ResultBottomSheet';

export default ResultBottomSheet;
