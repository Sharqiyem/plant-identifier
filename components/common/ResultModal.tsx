import React, { useState } from 'react';
import { View, Text, TouchableOpacity, Modal, FlatList, SafeAreaView, Image } from 'react-native';
import { PlantWithMeta, Language } from '@/types';
import { PlantCard } from '@/components/common/PlantCard';
import { Ionicons } from '@expo/vector-icons';
import Colors from '@/constants/Colors';

interface ResultModalProps {
  visible: boolean;
  onClose: () => void;
  plantInfo: { [key: string]: PlantWithMeta };
  languages: Language[];
}

export const ResultModal: React.FC<ResultModalProps> = ({
  visible,
  onClose,
  plantInfo,
  languages
}) => {
  const [selectedLanguage, setSelectedLanguage] = useState(languages[0]?.languageCode || 'en');

  const languageCodes = Object.keys(plantInfo);
  const previewUri = plantInfo[languageCodes[0]]?.previewUri;

  return (
    <Modal animationType="slide" transparent={true} visible={visible} onRequestClose={onClose}>
      <SafeAreaView className="flex-1 bg-black/70">
        <View className="flex-1 m-4 bg-card rounded-lg overflow-hidden">
          <View className="flex-row justify-between items-center p-4 border-b border-border">
            <Text className="text-2xl font-bold text-card-foreground">
              Plant Identification Results
            </Text>
            <TouchableOpacity
              testID="close-button"
              onPress={onClose}
              className="rounded-full p-2 bg-primary "
              accessibilityLabel="Close modal"
              hitSlop={{ top: 10, bottom: 10, left: 10, right: 10 }}
            >
              <Ionicons name="close" size={24} color={Colors.text} />
            </TouchableOpacity>
          </View>

          {previewUri && (
            <Image
              testID="plant-image"
              source={{ uri: previewUri }}
              className="w-full h-48 object-cover"
            />
          )}

          {plantInfo[selectedLanguage] && (
            <View className="px-4">
              <PlantCard testID="plant-card" plant={plantInfo[selectedLanguage]} />
            </View>
          )}

          <Text className="text-lg font-semibold text-primary mt-4 mb-2 px-4">All Languages</Text>

          <FlatList
            data={languageCodes}
            renderItem={({ item }) => (
              <View className="mb-6 px-4">
                <Text className="text-lg font-semibold text-accent mb-2">
                  {languages.find((lang) => lang.languageCode === item)?.languageName || item}
                </Text>
                <PlantCard testID="plant-card" plant={plantInfo[item]} />
              </View>
            )}
            keyExtractor={(item) => item}
          />
        </View>
      </SafeAreaView>
    </Modal>
  );
};
