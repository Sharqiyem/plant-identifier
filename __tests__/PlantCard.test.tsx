import React from 'react';
import { render } from '@testing-library/react-native';
import { PlantCard } from '@/components/common/PlantCard';

describe('PlantCard', () => {
  const mockPlant = {
    name: 'Mock Plant',
    scientificName: 'Mockus plantus',
    family: 'Mockaceae',
    description: 'This is a mock plant for testing purposes.'
  };

  it('renders correctly with plant information', () => {
    const { getByTestId } = render(<PlantCard plant={mockPlant} />);

    expect(getByTestId('plant-name')).toHaveTextContent('Name: Mock Plant');
    expect(getByTestId('plant-scientific-name')).toHaveTextContent(
      'Scientific Name: Mockus plantus'
    );
    expect(getByTestId('plant-family')).toHaveTextContent('Family: Mockaceae');
    expect(getByTestId('plant-description')).toHaveTextContent(
      'This is a mock plant for testing purposes.'
    );
  });

  it('does not render timestamp by default', () => {
    const { queryByTestId } = render(<PlantCard plant={mockPlant} />);

    expect(queryByTestId('plant-timestamp')).toBeNull();
  });

  it('renders timestamp when showTimestamp is true and timestamp is provided', () => {
    const timestamp = '2023-05-20T12:00:00Z';
    const { getByTestId } = render(
      <PlantCard plant={mockPlant} showTimestamp timestamp={timestamp} />
    );

    expect(getByTestId('plant-timestamp')).toBeTruthy();
    expect(getByTestId('plant-timestamp')).toHaveTextContent(new Date(timestamp).toLocaleString());
  });

  it('applies correct props', () => {
    const { getByTestId } = render(<PlantCard plant={mockPlant} />);

    const descriptionElement = getByTestId('plant-description');

    expect(descriptionElement).toHaveProp('numberOfLines', 3);
  });

  it(`renders match with snapshot`, () => {
    const tree = render(<PlantCard plant={mockPlant} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
