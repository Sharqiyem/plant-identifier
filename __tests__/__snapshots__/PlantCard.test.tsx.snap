// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantCard renders match with snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "hsl(221.2, 50%, 10%)",
      },
      {
        "paddingBottom": 16,
        "paddingTop": 16,
      },
      {
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
      },
      {
        "marginBottom": 16,
      },
      {
        "elevation": 6,
      },
      {
        "shadowColor": "rgba(0, 0, 0, 0.1)",
        "shadowOffset": {
          "height": 10,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 15,
      },
      {
        "shadowColor": "rgba(0, 0, 0, 0.1)",
        "shadowOffset": {
          "height": 10,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 15,
      },
      {
        "shadowColor": "rgba(0, 0, 0, 0.5)",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "hsl(221.2, 5%, 90%)",
        },
        {
          "fontSize": 24,
          "lineHeight": 32,
        },
        {
          "fontWeight": "700",
        },
        {
          "marginBottom": 8,
        },
      ]
    }
    testID="plant-name"
  >
    Name: 
    Mock Plant
  </Text>
  <Text
    style={
      [
        {
          "color": "hsl(221.2, 5%, 60%)",
        },
        {
          "marginBottom": 4,
        },
      ]
    }
    testID="plant-scientific-name"
  >
    Scientific Name: 
    Mockus plantus
  </Text>
  <Text
    style={
      [
        {
          "color": "hsl(221.2, 5%, 60%)",
        },
        {
          "marginBottom": 4,
        },
      ]
    }
    testID="plant-family"
  >
    Family: 
    Mockaceae
  </Text>
  <Text
    numberOfLines={3}
    style={
      [
        {
          "color": "hsl(221.2, 5%, 90%)",
        },
        {
          "marginBottom": 8,
        },
      ]
    }
    testID="plant-description"
  >
    This is a mock plant for testing purposes.
  </Text>
</View>
`;
